wled00/FX_2Dfcn.cpp

```diff
@@ -168,50 +168,63 @@ uint16_t IRAM_ATTR_YN Segment::XY(int x, int y)
   return isActive() ? (x%vW) + (y%vH) * vW : 0;
 }
 
+// raw setColor function without checks (checks are done in setPixelColorXY())
+void IRAM_ATTR_YN Segment::_setPixelColorXY_raw(int& x, int& y, uint32_t& col)
+{
+#ifndef WLED_DISABLE_MODE_BLEND
+  // if blending modes, blend with underlying pixel
+  if (_modeBlend) col = color_blend(strip.getPixelColorXY(start + x, startY + y), col, 0xFFFFU - progress(), true);
+#endif
+  strip.setPixelColorXY(start + x, startY + y, col);
+  if (mirror) { //set the corresponding horizontally mirrored pixel
+    if (transpose) strip.setPixelColorXY(start + x, startY + height() - y - 1, col);
+    else           strip.setPixelColorXY(start + width() - x - 1, startY + y, col);
+  }
+  if (mirror_y) { //set the corresponding vertically mirrored pixel
+    if (transpose) strip.setPixelColorXY(start + width() - x - 1, startY + y, col);
+    else           strip.setPixelColorXY(start + x, startY + height() - y - 1, col);
+  }
+  if (mirror_y && mirror) { //set the corresponding vertically AND horizontally mirrored pixel
+    strip.setPixelColorXY(start + width() - x - 1, startY + height() - y - 1, col);
+  }
+}
+
 void IRAM_ATTR_YN Segment::setPixelColorXY(int x, int y, uint32_t col)
 {
   if (!isActive()) return; // not active
 
   const int vW = vWidth();   // segment width in logical pixels (can be 0 if segment is inactive)
   const int vH = vHeight();  // segment height in logical pixels (is always >= 1)
-  if (unsigned(x) >= vW || unsigned(y) >= vH) return;  // if pixel would fall out of virtual segment just exit
+  if (unsigned(x) >= unsigned(vW) || unsigned(y) >= unsigned(vH)) return;  // if pixel would fall out of virtual segment just exit
 
   // if color is unscaled
   if (!_colorScaled) col = color_fade(col, _segBri);
 
   if (reverse  ) x = vW - x - 1;
   if (reverse_y) y = vH - y - 1;
   if (transpose) { std::swap(x,y); } // swap X & Y if segment transposed
-  x *= groupLength(); // expand to physical pixels
-  y *= groupLength(); // expand to physical pixels
-  int W = width();
-  int H = height();
-
-  int yY = y;
-  for (int j = 0; j < grouping; j++) {   // groupping vertically
-    if (yY >= H) break;
-    int xX = x;
-    for (int g = 0; g < grouping; g++) { // groupping horizontally
-      if (xX >= W) break;  // we have reached X dimension's end
-#ifndef WLED_DISABLE_MODE_BLEND
-      // if blending modes, blend with underlying pixel
-      if (_modeBlend) col = color_blend(strip.getPixelColorXY(start + xX, startY + yY), col, 0xFFFFU - progress(), true);
-#endif
-      strip.setPixelColorXY(start + xX, startY + yY, col);
-      if (mirror) { //set the corresponding horizontally mirrored pixel
-        if (transpose) strip.setPixelColorXY(start + xX, startY + height() - yY - 1, col);
-        else           strip.setPixelColorXY(start + width() - xX - 1, startY + yY, col);
+  unsigned groupLen = groupLength();
+
+  if(groupLen > 1)
```

Comment:
This reduces the processing if group length is 1 by removing:
- four assignments (`yY` && `xX`, `W` & `H`)
  - hidden two subtractions and a comparison (`W` & `H`)
- two multiplications
- two comparisons (`j` & `g`, `uint8_t`!)
- two comparisons (`yY>=H` && `xX>=W`)
- two increments (`xX` & `yY`)

For each `setPixelColorXY()` call.

What may seem irrelevant with low number of pixels actually produces a noticeable slowdown on large pixel count.